ubi_manifest/worker/tasks/depsolver/rpm_depsolver.py

```diff
@@ -207,19 +207,34 @@ def resolve_rpms(
         return self.what_provides(batch, "provides.name", repos, blacklist)
 
     def get_source_pkgs(
-        self, binary_rpms: list[UbiUnit], blacklist: list[PackageToExclude]
+        self,
+        binary_rpms: set[UbiUnit],
+        binary_repos: list[YumRepository],
+        blacklist: list[PackageToExclude],
     ) -> set[UbiUnit]:
-        crit = create_or_criteria(
-            ["filename"], [(rpm.sourcerpm,) for rpm in binary_rpms if rpm.sourcerpm]
-        )
-
+        srpm_repos = set()
+        filenames = set()
+
+        for repo in binary_repos:
+            if not repo.relative_url:
+                continue
+            for rpm in binary_rpms:
+                if rpm.sourcerpm and rpm.associate_source_repo_id == repo.id:
+                    srpm_repo = repo.get_source_repository().result()
```

Comment:
This is going to do too many unnecessary requests to pulp:
* `number of RPMs` * `number of binary repos`

You can count on following 1:1 mapping - binary repo : srpm repo and get these before any loop starts.
Then you need create some map of `srpm_repo:[srpm]`  so you can query in the expected repo and do required queries.